x-pack/plugins/ml/public/application/datavisualizer/stats_datagrid/stats_datagrid.tsx

```diff
@@ -0,0 +1,251 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { useMemo, useState } from 'react';
+
+import {
+  CENTER_ALIGNMENT,
+  EuiButtonIcon,
+  EuiFlexItem,
+  EuiIcon,
+  EuiInMemoryTable,
+  EuiText,
+  HorizontalAlignment,
+  LEFT_ALIGNMENT,
+  RIGHT_ALIGNMENT,
+} from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import { EuiTableComputedColumnType } from '@elastic/eui/src/components/basic_table/table_types';
+import { FieldTypeIcon } from '../../components/field_type_icon';
+import { FieldVisConfig } from '../index_based/common';
+import { DataVisualizerFieldExpandedRow } from './expanded_row';
+import { DocumentStat } from './components/field_data_row/document_stats';
+import { DistinctValues } from './components/field_data_row/distinct_values';
+import { NumberContentPreview } from './components/field_data_row/number_content_preview';
+import { DataVisualizerIndexBasedAppState } from '../../../../common/types/ml_url_generator';
+import { useTableSettings } from '../../data_frame_analytics/pages/analytics_management/components/analytics_list/use_table_settings';
+import { TopValuesPreview } from './components/field_data_row/top_values_preview';
+import type { MlJobFieldType } from '../../../../common/types/field_types';
+const FIELD_NAME = 'fieldName';
+
+interface DataVisualizerDataGrid {
+  items: FieldVisConfig[];
+  pageState: DataVisualizerIndexBasedAppState;
+  updatePageState: (update: Partial<DataVisualizerIndexBasedAppState>) => void;
+}
+
+export type ItemIdToExpandedRowMap = Record<string, JSX.Element>;
+
+function getItemIdToExpandedRowMap(
+  itemIds: string[],
+  items: FieldVisConfig[]
+): ItemIdToExpandedRowMap {
+  return itemIds.reduce((m: ItemIdToExpandedRowMap, fieldName: string) => {
+    const item = items.find((fieldVisConfig) => fieldVisConfig[FIELD_NAME] === fieldName);
+    if (item !== undefined) {
+      m[fieldName] = <DataVisualizerFieldExpandedRow item={item} />;
+    }
+    return m;
+  }, {} as ItemIdToExpandedRowMap);
+}
+
+export const DataVisualizerDataGrid = ({
+  items,
+  pageState,
+  updatePageState,
+}: DataVisualizerDataGrid) => {
+  const [expandedRowItemIds, setExpandedRowItemIds] = useState<string[]>([]);
+  const [expandAll, toggleExpandAll] = useState<boolean | undefined>();
+
+  const { onTableChange, pagination, sorting } = useTableSettings<FieldVisConfig>(
+    items,
+    pageState,
+    updatePageState
+  );
+  const showDistributions: boolean = pageState.showDistributions ?? true;
+  const toggleShowDistribution = () => {
+    updatePageState({
+      ...pageState,
+      showDistributions: !showDistributions,
+    });
+  };
+
+  function toggleDetails(item: FieldVisConfig) {
+    if (item.fieldName === undefined) return;
+    const index = expandedRowItemIds.indexOf(item.fieldName);
+    if (index !== -1) {
+      expandedRowItemIds.splice(index, 1);
+      setExpandedRowItemIds([...expandedRowItemIds]);
```

Comment:
you have `setExpandedRowItemIds` at the end of the function, so it called twice. I guess this one should be removed. 
```suggestion
```

Reply:
Updated here 4ea6d17